torch/distributed/_composable/fsdp/fully_shard.py

```diff
@@ -3,21 +3,30 @@
 import typing_extensions
 
 import torch.nn as nn
+from torch._prims_common import DeviceLikeType
 
 from torch.distributed._composable import contract
 from torch.distributed._composable_state import _insert_module_state
 
+from ._fsdp_init import _normalize_device
+from ._fsdp_state import FSDPState
 
-@contract()
+
+@contract(state_cls=FSDPState)
 def fully_shard(
     module: nn.Module,
+    *,
+    device: DeviceLikeType = "cuda",
 ):
     if isinstance(module, (nn.ModuleList, nn.ModuleDict)):
         raise ValueError(
             f"fully_shard does not support containers that do not implement forward: {module}"
         )
+    device = _normalize_device(device)
     state = fully_shard.state(module)
     _insert_module_state(module, state)
+    state._module = module
+    state._device = device
```

Comment:
nit: state init seems be scatterred vertically. is it possible to have ``fully_shard.state(*args, **kwargs)`` where we put can pack ``.module=`` ``_device`` logic into ``fully_shard.state(*args, **kwargs)`` ?

Reply:
Good point. I think the only constraint is that the state object is constructed in the `contract` decorator, so we cannot use the `FSDPState.__init__()` directly.

Talking out loud: Even if we have a separate `FSDPState.init()` function (distinct from `__init__()`), we will eventually have to have 1 line per assignment like `self._module = module` and `self._device = device`, etc. I am not sure if there is a way around that. What do you think?

Reply:
I feel a separate `FSDPState.init()` would be cleaner, and we can put the `_insert_module_state` to the state class itself? this makes relatively clean separation between state init and fsdp core logic

Reply:
Sounds good. I added an `FSDPState.init()`.